### PR TITLE
Refactoring services to use Interfaces instead of the BaseService

### LINKED ISSUE #429: Refactor the BaseService
Hi there,

As described in the Slack channel I have identified that the BaseService is no longer fit for purpose and propose to refactor this to split up the functionality in the services into relevant Interfaces that the services can then implement. This will support composition over inheritance and better serve the current services.

I'm happy to pick this one if you wish to assign it back to me please.

John

### PR DESCRIPTION
**Pull request by jhanna60**
## Summary of change

I have created 3 new interfaces to support the services Findable, Savable and Deletable, I have removed the Base Service and updated the Services to use the interfaces instead. I have also refactored the PredefinedShelfServiceTest to include the .count() call that I made in an earlier PR and removed the two redundant delete tests in here.

## Related issue

Closes #429 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-h6vx3n8l-KN8QnO50r7QWZHgHFnFdPw)

### CODE REVIEW

File: src/test/java/com/karankumar/bookproject/backend/service/PredefinedShelfServiceTest.java [hunk 1]
```diff
@@ -1,25 +1,6 @@
...
-    License, or (at your option) any later version.
-
-    This program is distributed in the hope that it will be useful, but WITHOUT ANY
-    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
-    PURPOSE.  See the GNU General Public License for more details.
-
-    You should have received a copy of the GNU General Public License along with this program.
-    If not, see <https://www.gnu.org/licenses/>.
- */
-
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/service/PredefinedShelfServiceTest.java:17]: Could you please revert this change?

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for implementing this, it's great! 

The only thing to do would be to attach our [copyright notice](https://github.com/Project-Books/book-project/blob/master/docs/COPYRIGHT) to the start of the new files and bring it back in `PredefinedShelfServiceTest`. Once this is done, this should be ready to be merged

File: src/test/java/com/karankumar/bookproject/backend/service/PredefinedShelfServiceTest.java (hunk 1 above)
Comment (jhanna60) [reply src/test/java/com/karankumar/bookproject/backend/service/PredefinedShelfServiceTest.java]: Hi @knjk04 sorry my IDE collapsed these and this one got removed, I've reinstated and also added it to the new Interface files, apologies for this.

Cheers
John

**Comment by amit1307** [comment]
Just a quick comment on this change, I agree having base service class was not a good idea but moving this to interface may also not be a great option due to the below reasons:

1. There is no good use case for interfaces as such in our app as of now. If you see these interfaces are not used anywhere so it doesn't add any value. If we were using the interfaces and under the hood implementation was changing a lot then may be it would be a better use case.
2. The service classes are looking very noisy with every service class implementing minimum 2 or 3 interfaces.
3. What's the disadvantages if the service classes don't extend or implement any interface?

Otherwise the PR looks good but just trying to understand if we really need this?

**Comment by jhanna60** [comment]
Hi @amit1307 its a valid reply overall and the changes were made off the back of noticing that BaseService isn't doing a good job as an Abstract class just now its forcing the delete functionality onto a service that doesn't need it.

We are looking at a couple of things I guess.. what was the reason for having that BaseService to start with? I assume it was to provide some standard structure for any service wishing to extend it? I have only built on this premise by breaking down the individual pieces of functionality that exist in there and having them as separate interfaces.

Some of the benefits I guess is that when working with multiple services we know if they are findable they will have a definition for findByID for example and the same for the other functionality provided by others. So we don't need to worry or concern ourselves about the concrete implementation inside each service only that if its type 'findable' it will have the relevant find methods internally.

When it comes to testing and mocking it makes it very easy to change implementations and inject them if we are using interfaces and decreases the reliance on external processes dealing with and knowing about concrete implementations.

The point about 'noise' in the code is valid but it can come down to personal preference this one, the trend towards Composition over inheritance is making this more and more common nowadays. The Interface Segregation principle in SOLID is being broken just now by the use of that BaseService class. It is forcing the predefinedShelfService to have a dud implementation of delete and deleteall and then to cover this fact there is some tests to make sure this functionality never works?

Happy to discuss further if you wish, or if you'd prefer me to refactor some more I'm always keen to learn I'm sure a good outcome can be achieved here.

I'm sure however you will also agree this is a bad code smell and needs to be dealt with? :

    @Override
    public void delete(PredefinedShelf shelf) {
        // Don't want to delete a predefined shelf
        LOGGER.severe("delete() should not be called");
    }

    @Override
    public void deleteAll() {
        // Don't want to delete the predefined shelves
        LOGGER.severe("deleteAll() should not be called");
    }

**Comment by amit1307** [comment]
> We are looking at a couple of things I guess.. what was the reason for having that BaseService to start with? I assume it was to provide some standard structure for any service wishing to extend it? I have only built on this premise by breaking down the individual pieces of functionality that exist in there and having them as separate interfaces.

Yes I completely agree with your first point where you mention the basis of this refactoring was the super BaseService that was enforcing an unwanted behaviour on child classes. So that itself tells not all the service classes need to be child of that one super base service class. 

> Some of the benefits I guess is that when working with multiple services we know if they are findable they will have a definition for findByID for example and the same for the other functionality provided by others. So we don't need to worry or concern ourselves about the concrete implementation inside each service only that if its type 'findable' it will have the relevant find methods internally

I am not sure about this benefit, for me its like if a method that I need already exists in the service I can use it and if not we can just create it. I personally wouldn't prefer to have the service implementing an interface just to help me what the service can contain, no?

>When it comes to testing and mocking it makes it very easy to change implementations and inject them if we are using interfaces and decreases the reliance on external processes dealing with and knowing about concrete implementations.

I agree to this point in general but not in the context of our app and that was one of main reasons I raised the questions in the first place. If by testing you mean the unit testing in that case we would most likely be invoking/mocking public methods and for that it doesn't really matter if we do that on the interface type or the actual object type.

>The point about 'noise' in the code is valid but it can come down to personal preference this one, the trend towards Composition over inheritance is making this more and more common nowadays. The Interface Segregation principle in SOLID is being broken just now by the use of that BaseService class. It is forcing the predefinedShelfService to have a dud implementation of delete and deleteall and then to cover this fact there is some tests to make sure this functionality never works?

Yes `BaseService` class was not helping us in a great way and was coupling all the services to same contract. 

>I'm sure however you will also agree this is a bad code smell and needs to be dealt with? :

Yes definitely.

Having said all that how would you and @knjk04 would feel if we completely remove `BaseService` and not replace it with anything. We don't really need to have our services extend or implement anything for now and it can always be revisited and changed if there is a proper use case?

**Comment by jhanna60** [comment]
Hi @amit1307 thanks for the reply, I think its a fair suggestion that we remove that base service. If @knjk04 is also in agreement I can create an issue and am more than happy to implement this?

Regards

**Comment by knjk04** [comment]
@jhanna60 That's fine, feel free to create a new issue for removing the base service and I'll assign you that. Apologies for the change in requirements. 

I'll close this PR and its corresponding issue
